### PR TITLE
Add CancellationError

### PR DESCRIPTION
**Pull request by mtsmfm**
#### What it does

Fixes: https://github.com/eclipse-theia/theia/issues/9988.

Add CancellationError to be compatible with VSCode.

https://github.com/eclipse-theia/theia/issues/9136#issuecomment-809660980

#### Review checklist

- [ ] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by vince-fugnitto** [comment]
@mtsmfm fixes https://github.com/eclipse-theia/theia/issues/9988?

**Comment by mtsmfm** [comment]
@vince-fugnitto Sorry I overlooked that issue. Yes, this PR fixes that.

**Comment by tsmaeder** [comment]
@mtsmfm thanks for the PR. How can I test it? Why I'm asking is this: what happens when a plugin actually throws "CancelledError"? I think we should handle that more gracefully than to just display an error. What does VS Code do in this instance?

**Comment by paul-marechal** [comment]
@tsmaeder VS Code made a function to intercept some calls and log or not some of the errors:

https://github.com/microsoft/vscode/blob/1.60.1/src/vs/workbench/api/common/extHostLanguageFeatures.ts#L1595-L1624

It is then extensively used, but they parameterize it in a few places to ignore logging `CancellationError`:

- https://github.com/microsoft/vscode/blob/1.60.1/src/vs/workbench/api/common/extHostLanguageFeatures.ts#L1927
- https://github.com/microsoft/vscode/blob/1.60.1/src/vs/workbench/api/common/extHostLanguageFeatures.ts#L1941

Since this is just related to logging, I think it is fine to do this as a follow-up.

**Comment by paul-marechal** [comment]
@mtsmfm overall the changes looks ok to me, but do you know which plugin/version I could use to compare the state of master vs your PR? I tried the latest release of `rust-analyzer` but I got a nasty path bug error on start, which I assume comes from our implementations...

**Comment by tsmaeder** [comment]
@mtsmfm Theia is handling errors on remote calls generally here: https://github.com/eclipse-theia/theia/blob/96930a2edcc4c706d41ab8486500de7979d01283/packages/core/src/common/messaging/proxy-factory.ts#L247

I agree we should not delay merging this PR because we're not "properly" handling `CancellationError` because of the following argument: for a plugin that does not use `CancellationError`, nothing changes. For a plugin that _does_ use cancellation error, we now would get a different error each time it tries to throw `CancellationError`. So in both cases, we end up with a log full of bogus error messages. So as far as I'm concerned, it passes the "the code is no worse than before" test.

**Comment by tsmaeder** [comment]
The code looks good to me, but I can't provoke the error from https://github.com/eclipse-theia/theia/issues/9136#issuecomment-809660980 on master. @vince-fugnitto you had analyzed the problem, right?

**Comment by vince-fugnitto** [comment]
> The code looks good to me, but I can't provoke the error from [#9136 (comment)](https://github.com/eclipse-theia/theia/issues/9136#issuecomment-809660980) on master. @vince-fugnitto you had analyzed the problem, right?

@tsmaeder I don't really remember, in the issue all I did was help identify where `CancellationError` came from, and how someone might implement it.

**Comment by mtsmfm** [comment]
@paul-marechal @tsmaeder Currently rust-analyzer compiles its extension for node 14, not 12. So latest rust-analyzer doesn't work because theia only supports 12 until https://github.com/eclipse-theia/theia/pull/9936

Can you try my fork? https://github.com/mtsmfm/rust-analyzer/releases/tag/nightly

It applies a patch to change target

https://github.com/mtsmfm/rust-analyzer/pull/1

---

I tried che-theia with my patch and with my fork (https://gist.github.com/mtsmfm/a9b34a4004d20c38caf8e06f7929feb4) and I confirmed `Class extends value undefined is not a constructor or null` is not appeared.
I got another error now though.

> Activating extension 'rust-analyzer (nightly)' failed: The language client requires VS Code version ^1.53.0 but received version 1.50.0

image

For che-theia, rust-analyzer still doesn't work since its dependency requires vscode ^1.53.0.

https://github.com/rust-analyzer/rust-analyzer/blob/f1d7f98ed07b9934286b9c4809dd4d7a47537879/editors/code/package-lock.json#L3764
https://github.com/microsoft/vscode-languageserver-node/blob/release/client/7.1.0-next.5/client/src/node/main.ts#L25

I think if you can use latest theia, it should work since theia's default is 1.53.2 now.
https://github.com/eclipse-theia/theia/pull/9959

For che-theia, I think this patch is needed
https://github.com/eclipse-che/che-theia/pull/1213

**Comment by mtsmfm** [comment]
Or perhaps this release can work without patch since it's the last release which supports node 12

https://github.com/rust-analyzer/rust-analyzer/releases/tag/2021-08-09

**Comment by tsmaeder** [comment]
Luckily, the newest vscode-java uses `CancellationError`.

File: packages/plugin-ext/src/plugin/quick-open.ts [hunk 1]
```diff
@@ -40,6 +40,13 @@ export function isPromiseCanceledError(error: any): boolean {
     return error instanceof Error && error.name === canceledName && error.message === canceledName;
 }
 
+export class CancellationError extends Error {
```
Comment (tsmaeder) [inline packages/plugin-ext/src/plugin/quick-open.ts:43]: I just noticed that this is in quick-open.ts. However, this is a general class and should be moved to `types-impl.ts`

**Review by tsmaeder — CHANGES_REQUESTED** [review]
I was able to test with vscode-java v.0.82.0, but we should move the type declaration.

File: packages/plugin-ext/src/plugin/quick-open.ts (hunk 1 above)
Comment (tsmaeder) [reply packages/plugin-ext/src/plugin/quick-open.ts:43]: Let me correct that: we're consuming cancellation-related stuff from core/common/cancellation.ts. We probably put the implementation of the class there, as well.

**Comment by mtsmfm** [comment]
@tsmaeder I've moved class into core/common/cancellation.ts. Can you review again?

**Review by tsmaeder — APPROVED** [review]
lgtm now

**Pull request merged**